[새리] 미션3: 체스판 초기화

sallyjellyy opened the pull request:
안녕하세요, 새리입니다. 미션3 체스판 초기화 코드리뷰 부탁드리겠습니다.

- Pawn.java
  - representation을 컬러와 같이 상수로 표현하였습니다.
  - 기본생성자에 representation을 추가하였고 기존에 컬러를 매개변수로 받던 생성자에 representation을 매개변수로 추가해주었습니다.
  - representation에 대한 getter 메소드도 추가하였습니다.
- Board.java
  - 기존에 있던 멤버변수 pawns 리스트가 whitePawns와 blackPawns로 바뀌었습니다.
  - 그에 따라 기존에 pawns에 대한 메소드들도 흰색과 검은색 둘로 나뉘게 되었습니다.
  - 과정에서 중복되는 코드가 있었지만 접근제어자가 프라이빗이라 크게 변경하지는 못했고 getPawnsResult 메서드의 내용을 묶어 하나의 프라이빗 메소드로 지정하였고 그 getWhitePawnsResult와 getBlackPawnsResult메서드는 묶인 메서드를 호출하는 방식으로 구현하였습니다.
  - 보드 클래스의 생성자를 추가해주었습니다. 멤버변수인 whitePawns와 blackPawns를 초기화시켜줍니다.
  - 보드사이즈를 상수로 지정해주었습니다.
  - Initialize() 메서드를 추가하였습니다. 보드사이즈(8)만큼 흰색 폰과 검정색 폰을 추가해줍니다.
- PawnTest.java
  - 처음에는 미션 요구사항에 맞게 먼저 테스트코드를 추가하였습니다.
  - Pawn 클래스를 수정하면서 그에 맞게 테스트코드도 수정하였습니다.
- BoardTest.java
  - 미션1 코드리뷰에서 받은 피드백을 토대로 assertThat을 여러개 사용하는 대신 assertAll을 사용해보았으나.. 저렇게 하는 게 맞는지는 자신이 없습니다. 피드백 주시면 감사하겠습니다.
  - 폰테스트와 마찬가지로 수정된 board 클래스에 맞게 기존의 테스트코드를 수정하였습니다.
  - print() 메서드를 확인하는 테스트코드를 추가하였습니다.
  - Initialize() 메서드를 확인하는 테스트코드를 미션 요구사항에 따라 추가하였습니다.

잘 부탁드립니다. 감사합니다!

sallyjellyy committed:
feat: Change PawnTest class to meet the Mission3 requirements

sallyjellyy committed:
feat: Create static final values of representations

- 출력 시 필요한 representation 값을 컬러의 상수값과 같이 상수를 만들어 주었습니다.

sallyjellyy committed:
feat: Create representation variable and change constructor

- 출력 시 필요한 representation을 멤버변수로 지정해주었고 represenation을 리턴해주는 getRepresentation() 메소드를 추가해주었습니다.
- 기존의 기본생성자와 color를 매개변수로 받는 생성자 모두 representation 값을 지정해줄 수 있도록 변경하였습니다.

sallyjellyy committed:
refactor: Fix Pawn creator in the setUp() method to adjusted creator.

- 변경된 폰 생성자에 맞게 setUp() 메소드 안의 폰 생성자를 representation을 매개변수로 받도록 수정하였습니다.

sallyjellyy committed:
feat: Add initialize() test method to BoardTest according to the requirement

sallyjellyy committed:
feat: Create initialize() method and getPawnsResult() method.

- To make the test code work initailize() method and getWhitePawnsResult(), getBlackPawnsResult() was needed in Board class.
- ArrayList whitePawns and blackPawns are more proper to differenciate colors instead of pawns(the one already exist).

sallyjellyy committed:
refactor: Refactor Board class

- getWhitePawnsResult() 메소드와 getBlackPawnsResult() 메소드에서 중복되는 부분을 getPawnsResult() 메소드로 묶어 중복을 최소화하였습니다.
- 기존에 있었던 멤버변수 pawns가 whitePawns와 blackPawns로 나누면서 pawns에 대한 메소드였던 addPawn(), getPawnSize(), findPawn() 메소드들을 white과 black으로 구분지었습니다.
- 보드사이즈는 항상 8이므로 보드사이즈를 가지는 상수 BOARD_SIZE를 만들고 initialize() 메서드에서 사용하였습니다.
.

sallyjellyy committed:
refactor: Refactor BoardTest.

- 변경된 보드클래스 메서드에 맞게 보드테스트에 있는 메서드들과 그에 맞는 매개변수를 받게 수정하였습니다.
- 여러개의 assertThat을 사용할 경우 assertAll을 사용하는 게 좋다는 피드백을 바탕으로 assertAll을 사용해보았습니다.

sallyjellyy committed:
feat: Add print() method.

- 보드의 현황을 출력해주는 print() 메서드를 추가하였습니다.
- 미션 요구사항에 맞게 StringBuilder를 사용하고 pawnsResult 메소드들을 사용하였습니다.

sallyjellyy committed:
feat: Create Game class

- Game class has main method that runs whole game.
- The main method takes user's command with Scanner.

sonypark commented on src/main/java/net/sally/chess/Board.java:
```diff
@@ -6,17 +6,74 @@
-        pawns.add(pawn);
+    public static int BOARD_SIZE = 8;
+
+    public Board() {
+        whitePawns = new ArrayList<>();
+        blackPawns = new ArrayList<>();
+    }
```
이런 경우는 기본생성자에서 초기화하기 보다 필드에서 초기화 해주는 게 나을 것 같습니다. 기본 생성자에서 ArrayList()를 초기화하는 로직만 있는데, 이렇게 되면 불필요하게 코드의 양이 늘어나고 가독성도 더 떨어지기 때문입니다.

sallyjellyy replied:
말씀해주신대로 수정하였습니다!

sonypark commented on src/main/java/net/sally/chess/Board.java:
```diff
@@ -6,17 +6,74 @@
+        return pawnsResult.toString();
+    }
+
+    public void addWhitePawn(Pawn pawn) {
+        whitePawns.add(pawn);
+    }
+
+    public void addBlackPawn(Pawn pawn) {
+        blackPawns.add(pawn);
+    }
```
특정 컬러의 pawn만 들어온다는 걸 보장할 수 있을까요? 서로 다른 컬러의 pawn이 들어올 경우의 방어로직도 고려해주세요

sallyjellyy replied:
조건문을 써서 한 번 고쳐보았는데 한 번 확인해주실 수 있을까요? 이런 식으로 하는 게 맞는 지 잘 모르겠습니다.

sonypark commented on src/main/java/net/sally/chess/Board.java:
```diff
@@ -6,17 +6,74 @@
+        return blackPawns.get(i);
+    }
+
+    public void print() {
+        StringBuilder printer = new StringBuilder();
+        printer.append("........\n");
+        printer.append(getBlackPawnsResult() + "\n");
+        printer.append("........\n");
+        printer.append("........\n");
+        printer.append("........\n");
+        printer.append("........\n");
+        printer.append(getWhitePawnsResult() + "\n");
+        printer.append("........");
+
+        System.out.println(printer.toString());
     }
```
출력 로직이 Board가 수행해야 할 역할인지도 고민해보세요. 
입출력 로직을 담당하는 클래스를 만드는 방법도 있습니다.
지금 당장 분리할 필요는 없고, 필요하다고 느끼는 시점에 분리해도 좋아요!

추가적으로 `\n` 대신 `lineSeparator` 를 사용하는 방식도 공부해보세요

sallyjellyy replied:
네 추후에 분리하겠습니다. 말씀해주신대로 lineSeparator 써봤습니다! 근데 이게 맞나요..?😅  그리고 "........\n"을 따로 상수로 빼서 쓰는 게 더 효율적일까요?

sonypark commented on src/main/java/net/sally/chess/Game.java:
```diff
@@ -0,0 +1,31 @@
+                    break;
+            }
+        }
+
+        sc.close();
```
👍

sonypark commented on src/main/java/net/sally/chess/Game.java:
```diff
@@ -0,0 +1,31 @@
+            }
+        }
+
+        sc.close();
```
자바 상수를 사용해보세요~

sallyjellyy replied:
수정해보았는데 이 부분도 확인 부탁드립니다! 자신이 없어요..

sonypark commented on src/main/java/net/sally/pieces/Pawn.java:
```diff
@@ -4,17 +4,27 @@
     }
 
-    public Pawn(String color) {
+    public Pawn(String color, char representation) {
         this.color = color;
+        this.representation = representation;
     }
```
👍

sonypark commented on src/test/java/net/sally/chess/BoardTest.java:
```diff
@@ -12,20 +14,45 @@
     @BeforeEach
     private void setUp() {
```
`@BeforEach` 메서드의 접근제어자는 private으로 하면 안 됩니다. default로 해주세요.
**Method Signatures**를 보면 아래와 같이 나와있습니다.
```
@BeforeEach methods must have a void return type, must not be private, and must not be static.
```
분명 private으로 하면 안된다고 나와있는데, 정상적으로 작동합니다. 그건 리플렉션이라는 기능 때문인데요. 스프링 관련 내용이니 지금은 그냥 넘어가셔도 됩니다. 궁금하시면 [여기](https://sowhat4.tistory.com/73) 한 번 읽어보세요🧐

sallyjellyy replied:
넵 피드백 감사합니다! 리플렉션은 깡패군요..

sonypark commented on src/test/java/net/sally/chess/BoardTest.java:
```diff
@@ -12,20 +14,45 @@
+
+    @Test
+    @DisplayName("Board initializes properly")
+    void initialize() throws Exception {
+        board.initialize();
```
`throws Exception` 이 필요한가요?

sonypark replied:
DisplayName을 한글/영어 중 하나로 통일해주세요~
영어로 적을거라면 메서드 이름으로 만들면 돼서 굳이 없어도 될 것 같네요

sallyjellyy replied:
DisplayName도 수정하였습니다! 익셉션은 사실 미션 가이드에 나와있는 테스트코드를 가져오면서 생각없이 넣었어요.. 소니 피드백을 받고 보니 딱히 필요가 없겠네요. 수정하겠습니다.

sonypark commented on src/test/java/net/sally/chess/BoardTest.java:
```diff
@@ -12,20 +14,45 @@
+        );
+    }
+
+    private void verifyBlackPawns(Board board, Pawn pawn, int index) {
+        board.addBlackPawn(pawn);
+        assertAll(
+                () -> assertThat(board.getBlackPawnsSize()).isEqualTo(index + 1),
+                () -> assertThat(board.findBlackPawn(index)).isEqualTo(pawn)
+        );
```
네이밍을 index보다는 size로 하는건 어떤가요?

sallyjellyy replied:
네 사이즈가 더 맞는 이름이겠네요! 고쳤습니다 :)

sonypark reviewed (changes requested):
안녕하세요 새리, 리뷰어 소니 입니다.😊
요구사항 대로 아주 잘 구현해주셨습니다.👍
지난번보다 더 발전하신게 눈에 보이네요!💯
조금 더 개선이 필요한 부분에 대해 코멘트 남겼으니 확인 부탁드릴게요!

sonypark commented on src/test/java/net/sally/chess/BoardTest.java:
```diff
@@ -12,20 +14,45 @@
+        verifyBlackPawns(board, blackPawn, 0);
+    }
+
+    private void verifyWhitePawns(Board board, Pawn pawn, int index) {
+        board.addWhitePawn(pawn);
+        assertAll(
+                () -> assertThat(board.getWhitePawnSize()).isEqualTo(index + 1),
+                () -> assertThat(board.findWhitePawn(index)).isEqualTo(pawn)
+        );
+    }
+
+    private void verifyBlackPawns(Board board, Pawn pawn, int index) {
+        board.addBlackPawn(pawn);
+        assertAll(
+                () -> assertThat(board.getBlackPawnsSize()).isEqualTo(index + 1),
+                () -> assertThat(board.findBlackPawn(index)).isEqualTo(pawn)
+        );
```
👍

sallyjellyy committed:
refactor: Refactor Board class

- 불필요한 생성자는 삭제하고 멤버변수는 필드에서 초기화해주었습니다.
- addWhitePawn() 메서드와 addBlackPawn() 메서드에 각각의 색의 pawn이 들어오는지 확인하는 조건문을 추가하였습니다.
- print() 메서드에 lineSeparator를 사용하였습니다.

sallyjellyy committed:
refactor: Refactor BoardTest

- verifyWhitePawns/verifyBlackPawns 메서드에 매개변수 네이밍을 index에서 size로 수정하였습니다.
- initialize() 테스트 메서드의 불필요한 exception을 지웠습니다.
- BeforeEach 어노테이션이 붙은 setUp() 메서드를 default 로 바꿨습니다.
- DisplayName의 내용을 한글로 통일하였습니다.

sallyjellyy committed:
refactor: Refactor Game class.

- 게임이 실행되는 while문 안의 swich문에서 커맨드를 상수로 받을 수 있도록 하였습니다.

sonypark commented on src/main/java/net/sally/chess/Board.java:
```diff
@@ -40,11 +35,15 @@ private String getPawnsResult(List<Pawn> pawns) {
-        blackPawns.add(pawn);
+        if (pawn.getColor().equals(Pawn.BLACK)) {
+            blackPawns.add(pawn);
+        }
```
조건문 안에 있는 로직을 Pawn 내부 메서드로 만들수도 있겠네요.  (ex.`isWhite()`, `isBlack()` )

sallyjellyy replied:
참고하여 수정했습니다. 감사합니다!

sonypark commented on src/main/java/net/sally/chess/Board.java:
```diff
@@ -6,17 +6,77 @@
+    }
+
+    public void print() {
+        StringBuilder printer = new StringBuilder();
+        printer.append("........\n");
+        printer.append("........\n");
+        printer.append(getBlackPawnsResult());
+        printer.append(System.lineSeparator());
+        printer.append("........\n");
+        printer.append("........\n");
+        printer.append("........\n");
+        printer.append("........\n");
+        printer.append("........\n");
+        printer.append(getWhitePawnsResult());
+        printer.append(System.lineSeparator());
+        printer.append("........");
```
```suggestion
        StringBuilder printer = new StringBuilder();
        printer.append("........" + NEW_LINE);
        printer.append("........" + NEW_LINE);
        printer.append(getBlackPawnsResult());
        printer.append(NEW_LINE);
        printer.append("........" + NEW_LINE);
        printer.append("........" + NEW_LINE);
        printer.append("........" + NEW_LINE);
        printer.append("........" + NEW_LINE);
        printer.append(getWhitePawnsResult());
        printer.append(NEW_LINE);
        printer.append("........");
```
이렇게 상수로 선언해놓고 사용하는 방식을 의도했었습니다 ㅎㅎ
`private static final String NEW_LINE = System.lineSeparator();`

"....." 같은 경우도 현재로서는 변하지 않는 문자열 값이니 상수로 처리해서 사용하는게 좋겠죠?
"....."과 NEW_LINE을 합쳐서 상수로 선언해도 되겠네요:)

sallyjellyy replied:
앗... 사실 처음에 그렇게 상수로 지정했다가.. lineSeparator를 검색하다 이렇게 되었어요..🥲 다시 수정하였습니다!

sonypark commented on src/test/java/net/sally/chess/BoardTest.java:
```diff
@@ -10,22 +12,46 @@
 
     @Test
     @DisplayName("폰이 보드에 정상적으로 등록")
     void create() throws Exception {
```
`Exception` 이 또 보이네요 🥲

sallyjellyy replied:
ㅠㅡㅠ또 있네요 수정하였습니다!

sallyjellyy replied:
소니 좋은 말씀 너무 감사해요. 제 코드를 누군가에게 보여주고 피드백을 받는 과정 자체가 익숙하지 않아서 틀리면 어쩌지 하는 생각을 항상 가지고 있는 것 같아요. 소니 말씀대로 좀 더 자신감 가져보도록 해보겠습니다. 감사합니다!

sonypark reviewed (changes requested):
변경사항 중 추가적으로 개선할 부분에 대해 한 번 더 코멘트 남겼습니다.
확인 부탁해요!😃
아 그리고 자신 없어 하시는 모습을 보이셔서 덧붙이자면, 눈치보지 말고 자신있게 코딩하세요! 😄
프로그래밍이라는게 딱 하나의 정답만 있는게 아니고, 여러 가지 방법 중에서 조금 더 나은 방법을 선택하는 것 뿐이거든요!
그리고 당연한 말이지만 저도 틀릴 수 있습니다. 제가 말씀드리는 내용도 항상 의심하시고 정말 이렇게 하는게 더 나은 방법인지, 더 좋은 방법은 없는지 찾아보고 공부해주세요. (리뷰어의 말을 있는 그대로 믿지 마세요..ㅋㅋㅋ)
새리가 작성하신 코드 모두 맞는 방법입니다. (코드가 원하는대로 동작하면 잘 짜신거에요!) 다만 리뷰를 통해 여기서 더 개선하면 좋을 점을 찾는 과정이니 부담 갖지 마시고 자유롭게, 자신있게 코딩하시면 좋겠어요! 잘 하고 계십니다👍👍

sallyjellyy committed:
refactor: Refactor Board class.

- 보드 출력 시에 필요한 BLANK를 상수로 두어 print() 메서드에서 쓸 수 있게 하였습니다.
- 폰을 추가해주는 메서드인 addWhitePawn/addBlackPawn에서 조건식을 pawn에서 새로 추가한 isWhite/isBlack을 사용하였습니다.

sallyjellyy committed:
refactor: Refactor Pawn class.

- Board 클래스에서 사용될, 폰 인스턴스의 색을 확인해주는 isWhite/isBlack 메서드를 추가하였습니다.

sallyjellyy committed:
refactor: Refactor BoardTest.

- create() 메서드의 불필요한 exception을 제거해주었습니다.

sonypark commented on src/main/java/net/sally/chess/Board.java:
```diff
@@ -6,17 +6,78 @@
-    public void addPawn(Pawn pawn) {
-        pawns.add(pawn);
+    public static int BOARD_SIZE = 8;
+    private static final String BLANK = "........\n";
+
```
```suggestion
    private static final String NEW_LINE = System.lineSeparator();
    private static final String BLANK = "........" + NEW_LINE;
```
이렇게 할수도 있겠네요 :)

sallyjellyy replied:
넵 말씀하신대로 수정했습니다.

sonypark commented on src/main/java/net/sally/chess/Board.java:
```diff
@@ -6,17 +6,78 @@
+        return blackPawns.get(i);
+    }
+
+    public void print() {
+        StringBuilder printer = new StringBuilder();
+        printer.append(BLANK)
+                .append(BLANK)
+                .append(getBlackPawnsResult())
+                .append(System.lineSeparator())
+                .append(BLANK)
+                .append(BLANK)
+                .append(BLANK)
+                .append(BLANK)
+                .append(BLANK)
+                .append(getWhitePawnsResult())
+                .append(System.lineSeparator())
+                .append(BLANK);
+
```
체스판 출력 결과를 확인해주세요~
출력이 잘못되고 있는 부분이 있네요🧐

sallyjellyy replied:
앗 그러네요😅 확인하고 수정했습니다.

sonypark commented on src/main/java/net/sally/pieces/Pawn.java:
```diff
@@ -4,17 +4,35 @@
+        return representation;
+    }
+
+    public boolean isWhite(Pawn pawn) {
+        return pawn.getColor().equals(Pawn.WHITE);
+    }
+
+    public boolean isBlack(Pawn pawn) {
+        return pawn.getColor().equals(Pawn.BLACK);
+    }
```
👍
WHITE는 Pawn 클래스에 정의된 상수이니 `Pawn.WHITE`가 아닌 Pawn 클래스 내부에서는 WHITE로 사용해도 된답니다~

sallyjellyy replied:
참고하여 수정하였습니다. 감사합니다!

sonypark reviewed (changes requested):
체스판 출력하는 부분 한 번 확인해주세요~

sallyjellyy committed:
refactor: Refactor Pawn class

- Changed Pawn.WHITE/Pawn.BLACK to WHITE/BLACK since putting pawn. is not necessary within the same class.

sallyjellyy committed:
refactor: Refactor Board.

- Created String constants for a new line and blank line for print() method.
- Fixed print() method to print properly(to meet the requirements)

sonypark reviewed (approved):
수고하셨습니다!
다음 단계 진행해주세요~💪
